### PR TITLE
Routing + Upgrade integration test for precaching.

### PR DESCRIPTION
**Pull request by gauntface**
R: @jeffposnick @addyosmani @philipwalton 

Hopefully this is representative of a service worker flow for precaching.

### CODE REVIEW

File: test/workbox-precaching/integration/precache-and-update.js [hunk 1]
```diff
@@ -22,34 +24,78 @@ describe(`[workbox-precaching] Precache and Update`, function() {
...
-  it(`should load a page with service worker `, async function() {
-    const testing = `${testServerAddress}/test/workbox-precaching/static/precache-and-update/`;
-    await webdriver.get(testing);
-    await webdriver.wait(() => {
-      return webdriver.executeAsyncScript((cb) => {
-        navigator.serviceWorker.getRegistration()
-        .then((registration) => {
-          if (!registration) {
-            return cb(false);
+  const activateSW = async (swFile) => {
```
[the remark this answers is not in the document]
Comment (gauntface) [reply test/workbox-precaching/integration/precache-and-update.js]: I'll take another pass at this.

Firefox and Chrome behaviour differ here but it's extremely difficult to diagnose what is SW issue vs WebDriver issue.

[the remark this answers is not in the document]
Comment (gauntface) [reply test/workbox-precaching/integration/precache-and-update.js]: Just tried your suggestions and they aren't working.

Firefox seems to have an issue with .ready when you unregister one sw and register a second one.

Chrome doesn't seem to resolve with the polymer snippet.

[the remark this answers is not in the document]
Comment (jeffposnick) [reply test/workbox-precaching/integration/precache-and-update.js]: Okay, something else to try: call `navigator.serviceWorker.register()` multiple times on the same page, with a different `scriptURL` for each. (If you want to use the same underlying script, just use `?step=1` and `?step=2`.)

Then set up a promise that resolves when there's a `controllerchange` event and the current SW controlling the page has the URL you expect.

```js
const secondScriptUrl = 'sw.js?step=2';
const secondSWIsInControl = new Promise((resolve) => {
  navigator.serviceWorker.addEventListener('controllerchange', () => {
    if (navigator.serviceWorker.controller.scriptURL === secondScriptUrl) {
      resolve();
    }
  });
});
navigator.serviceWorker.register(secondScriptUrl);

// Later:
await secondSWIsInControl;
```

[the remark this answers is not in the document]
Comment (jeffposnick) [reply test/workbox-precaching/integration/precache-and-update.js]: (If you're happy with the current code in this PR, i.e. it works in FF and Chrome, then we can just go ahead with that. It's just that this general area is fraught with race conditions, and it historically has helped to explicitly wait for the exact conditions that you need as a defense against flakiness.)

**Comment by gauntface** [comment]
@jeffposnick another review please.

File: test/workbox-precaching/integration/precache-and-update.js [hunk 2]
```diff
@@ -22,34 +24,54 @@ describe(`[workbox-precaching] Precache and Update`, function() {
...
+    const error = await webdriver.executeAsyncScript((swFile, cb) => {
+      navigator.serviceWorker.register(swFile)
+      .then(() => {
+        navigator.serviceWorker.addEventListener('controllerchange', () => {
+          if (navigator.serviceWorker.controller.scriptURL.endsWith(swFile)) {
```
Comment (jeffposnick) [inline test/workbox-precaching/integration/precache-and-update.js]: ❤️

**Pull request merged**